keyserver/server.go

```diff
@@ -0,0 +1,188 @@
+package keyserver
+
+import (
+	"crypto/tls"
+	"fmt"
+	"io/ioutil"
+	"log"
+	"os"
+	"os/signal"
+	"sync"
+	"syscall"
+	"time"
+
+	"github.com/BurntSushi/toml"
+	"github.com/coniks-sys/coniks-go/crypto/sign"
+	"github.com/coniks-sys/coniks-go/crypto/vrf"
+	"github.com/coniks-sys/coniks-go/merkletree"
+	"github.com/coniks-sys/coniks-go/protocol"
+	"github.com/coniks-sys/coniks-go/storage/kv"
+	"github.com/coniks-sys/coniks-go/storage/kv/leveldbkv"
+)
+
+type ServerConfig struct {
+	configFilePath      string
+	DatabasePath        string          `toml:"database"`
+	LoadedHistoryLength uint64          `toml:"loaded_history_length"`
+	Policies            *ServerPolicies `toml:"policies"`
+	TLS                 *TLSConnection  `toml:"tls"`
+}
+
+type TLSConnection struct {
+	PublicAddress string `toml:"public_address"` // address:port
+	LocalAddress  string `toml:"local_address"`
+	TLSCertPath   string `toml:"cert"`
+	TLSKeyPath    string `toml:"key"`
+}
+
+type ServerPolicies struct {
+	EpochDeadline merkletree.TimeStamp `toml:"epoch_deadline"`
+	VRFKeyPath    string               `toml:"vrf_key_path"`
+	SignKeyPath   string               `toml:"sign_key_path"` // it should be a part of policies, see #47
+	vrfKey        vrf.PrivateKey
+	signKey       sign.PrivateKey
+}
+
+type ConiksServer struct {
+	sync.RWMutex
+	dir *protocol.ConiksDirectory
+
+	stop     chan struct{}
+	waitStop sync.WaitGroup
+
+	db kv.DB // TODO: it is a placeholer for issue #37
+
+	configFilePath string
+	reloadChan     chan os.Signal
+	epochTimer     *time.Timer
+}
+
+func LoadServerConfig(path string) (*ServerConfig, error) {
+	var conf ServerConfig
+	if _, err := toml.DecodeFile(path, &conf); err != nil {
+		return nil, fmt.Errorf("Failed to load config: %v", err)
+	}
+
+	// load signing key
+	signKey, err := ioutil.ReadFile(conf.Policies.SignKeyPath)
```

Comment:
How does that reconcile with,

> Otherwise, -dir should not apply to generating the keys.

I'm going to push back and say that everything (keys, config, etc.) should be relative to `dir` and that the default for `dir` is `.`

That gives you this,

```
mkdir coniks/
coniksserver -dir coniks -genkey -genconfig
coniksserver -dir coniks  // run
```

but it also gives you,

```
coniksserver -genkey -genconfig
coniksserver
```

which seems to be what you want (the default being the current working directory, and not the path to the bin itself).

With the code I supplied above, you can always put an absolute path in the `config.toml` for various files.